File: Presentation/FormattingToolbar.swift
import Cocoa

@available(macOS 14.0, *)
extension InterviewMasterDelegate {
    // MARK: - Formatting Toolbar
    func setupFormattingToolbar(in parentView: NSView) {
        // Floating toolbar container - visionOS style
        formattingToolbar = NSVisualEffectView(frame: NSRect(
            x: (parentView.frame.width - 360) / 2,
            y: parentView.frame.height - 60,
            width: 360,
            height: 44
        ))
        formattingToolbar.autoresizingMask = [.minXMargin, .maxXMargin, .minYMargin]
        formattingToolbar.blendingMode = .withinWindow
        formattingToolbar.material = .hudWindow
        formattingToolbar.state = .active
        formattingToolbar.alphaValue = 0
        formattingToolbar.isHidden = true
        formattingToolbar.wantsLayer = true
        formattingToolbar.layer?.cornerRadius = 12
        formattingToolbar.layer?.borderWidth = 1.5
        formattingToolbar.layer?.borderColor = NSColor.white.withAlphaComponent(0.3).cgColor
        parentView.addSubview(formattingToolbar)

        let buttonWidth: CGFloat = 40
        let spacing: CGFloat = 4
        var xOffset: CGFloat = 8

        // Heading button
        let headingBtn = createFormattingButton(
            icon: "textformat.size",
            tooltip: "Heading",
            action: #selector(insertHeading)
        )
        headingBtn.frame.origin = NSPoint(x: xOffset, y: 7)
        formattingToolbar.addSubview(headingBtn)
        xOffset += buttonWidth + spacing

        // Bold button
        let boldBtn = createFormattingButton(
            icon: "bold",
            tooltip: "Bold",
            action: #selector(insertBold)
        )
        boldBtn.frame.origin = NSPoint(x: xOffset, y: 7)
        formattingToolbar.addSubview(boldBtn)
        xOffset += buttonWidth + spacing

        // Italic button
        let italicBtn = createFormattingButton(
            icon: "italic",
            tooltip: "Italic",
            action: #selector(insertItalic)
        )
        italicBtn.frame.origin = NSPoint(x: xOffset, y: 7)
        formattingToolbar.addSubview(italicBtn)
        xOffset += buttonWidth + spacing

        // Code button
        let codeBtn = createFormattingButton(
            icon: "chevron.left.forwardslash.chevron.right",
            tooltip: "Code",
            action: #selector(insertCode)
        )
        codeBtn.frame.origin = NSPoint(x: xOffset, y: 7)
        formattingToolbar.addSubview(codeBtn)
        xOffset += buttonWidth + spacing

        // Code block button
        let codeBlockBtn = createFormattingButton(
            icon: "curlybraces",
            tooltip: "Code Block",
            action: #selector(insertCodeBlock)
        )
        codeBlockBtn.frame.origin = NSPoint(x: xOffset, y: 7)
        formattingToolbar.addSubview(codeBlockBtn)
        xOffset += buttonWidth + spacing

        // List button
        let listBtn = createFormattingButton(
            icon: "list.bullet",
            tooltip: "Bullet List",
            action: #selector(insertList)
        )
        listBtn.frame.origin = NSPoint(x: xOffset, y: 7)
        formattingToolbar.addSubview(listBtn)
        xOffset += buttonWidth + spacing

        // Link button
        let linkBtn = createFormattingButton(
            icon: "link",
            tooltip: "Link",
            action: #selector(insertLink)
        )
        linkBtn.frame.origin = NSPoint(x: xOffset, y: 7)
        formattingToolbar.addSubview(linkBtn)
        xOffset += buttonWidth + spacing

        // Divider button
        let dividerBtn = createFormattingButton(
            icon: "minus",
            tooltip: "Divider",
            action: #selector(insertDivider)
        )
        dividerBtn.frame.origin = NSPoint(x: xOffset, y: 7)
        formattingToolbar.addSubview(dividerBtn)
    }

    func createFormattingButton(icon: String, tooltip: String, action: Selector) -> NSButton {
        let button = NSButton(frame: NSRect(x: 0, y: 0, width: 40, height: 30))
        button.image = NSImage(systemSymbolName: icon, accessibilityDescription: tooltip)
        button.bezelStyle = .rounded
        button.isBordered = false
        button.imagePosition = .imageOnly
        button.symbolConfiguration = NSImage.SymbolConfiguration(pointSize: 14, weight: .medium)
        button.target = self
        button.action = action
        button.toolTip = tooltip
        button.wantsLayer = true
        button.layer?.cornerRadius = 6
        button.contentTintColor = .white
        return button
    }

    func showFormattingToolbar() {
        guard let toolbar = formattingToolbar, !isFormattingToolbarVisible else { return }
        isFormattingToolbarVisible = true

        toolbar.isHidden = false
        NSAnimationContext.runAnimationGroup({ context in
            context.duration = 0.2
            toolbar.animator().alphaValue = 0.95
        })
    }

    func hideFormattingToolbar() {
        guard let toolbar = formattingToolbar, isFormattingToolbarVisible else { return }
        isFormattingToolbarVisible = false

        NSAnimationContext.runAnimationGroup({ context in
            context.duration = 0.2
            toolbar.animator().alphaValue = 0
        }, completionHandler: {
            toolbar.isHidden = true
        })
    }

    // MARK: - Formatting Actions
    @objc func insertHeading() {
        insertMarkdownWrapper(prefix: "## ", suffix: "", placeholder: "Heading")
    }

    @objc func insertBold() {
        insertMarkdownWrapper(prefix: "**", suffix: "**", placeholder: "bold text")
    }

    @objc func insertItalic() {
        insertMarkdownWrapper(prefix: "*", suffix: "*", placeholder: "italic text")
    }

    @objc func insertCode() {
        insertMarkdownWrapper(prefix: "`", suffix: "`", placeholder: "code")
    }

    @objc func insertCodeBlock() {
        insertMarkdownWrapper(prefix: "```\n", suffix: "\n```", placeholder: "code block")
    }

    @objc func insertList() {
        let range = textView.selectedRange()
        let selectedText = (textView.string as NSString).substring(with: range)

        guard !selectedText.isEmpty else { return }

        // Split by lines and add bullet to each line
        let lines = selectedText.components(separatedBy: .newlines)
        let bulletedText = lines.map { line in
            line.isEmpty ? "" : "- \(line)"
        }.joined(separator: "\n")

        textView.insertText(bulletedText, replacementRange: range)

        DispatchQueue.main.async {
            self.renderMarkdown()
        }
    }

    @objc func insertLink() {
        insertMarkdownWrapper(prefix: "[", suffix: "](url)", placeholder: "link text")
    }

    @objc func insertDivider() {
        let range = textView.selectedRange()
        textView.insertText("\n\n---\n\n", replacementRange: range)
        renderMarkdown()
    }

    func insertMarkdownWrapper(prefix: String, suffix: String, placeholder: String) {
        let range = textView.selectedRange()
        let selectedText = (textView.string as NSString).substring(with: range)

        // Only format if there's selected text
        guard !selectedText.isEmpty else { return }

        let textToInsert = prefix + selectedText + suffix
        textView.insertText(textToInsert, replacementRange: range)

        // Force re-render markdown
        DispatchQueue.main.async { [weak self] in
            self?.renderMarkdown()
        }
    }
}

File: Presentation/FormattingToolbar.cs
using System;
using System.Linq;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Media;
using System.Windows.Media.Animation;
using System.Windows.Threading;
using Microsoft.Windows.Themes;

namespace InterviewMasterApp.Presentation
{
    /// <summary>
    /// Floating formatting toolbar for markdown text editing.
    /// Ported from Swift: FormattingToolbar.swift
    /// Provides quick access to markdown formatting commands (heading, bold, italic, code, lists, links).
    /// </summary>
    public class FormattingToolbar
    {
        private Border? _toolbarContainer;
        private bool _isVisible;
        private TextBox? _textBox;
        private Action? _renderMarkdownCallback;

        /// <summary>
        /// Setup the formatting toolbar in the parent container.
        /// </summary>
        public void SetupFormattingToolbar(Panel parentView, Control textBox, Action renderMarkdownCallback)
        {
            _textBox = textBox as TextBox; // Only TextBox supports formatting operations
            _renderMarkdownCallback = renderMarkdownCallback;

            // Floating toolbar container - similar to visionOS style
            _toolbarContainer = new Border
            {
                Width = 360,
                Height = 44,
                HorizontalAlignment = HorizontalAlignment.Center,
                VerticalAlignment = VerticalAlignment.Top,
                Margin = new Thickness(0, 60, 0, 0),
                CornerRadius = new CornerRadius(12),
                BorderThickness = new Thickness(1.5),
                BorderBrush = new SolidColorBrush(Color.FromArgb(77, 255, 255, 255)), // 30% opacity white
                Background = new SolidColorBrush(Color.FromArgb(160, 30, 30, 30)), // Semi-transparent dark
                Opacity = 0,
                Visibility = Visibility.Hidden
            };

            // Use a StackPanel for buttons
            var buttonPanel = new StackPanel
            {
                Orientation = Orientation.Horizontal,
                Margin = new Thickness(8, 7, 8, 7),
                HorizontalAlignment = HorizontalAlignment.Center
            };

            const double spacing = 4;

            // Create formatting buttons
            buttonPanel.Children.Add(CreateFormattingButton("A", "Heading (##)", InsertHeading, spacing));
            buttonPanel.Children.Add(CreateFormattingButton("B", "Bold (**)", InsertBold, spacing));
            buttonPanel.Children.Add(CreateFormattingButton("I", "Italic (*)", InsertItalic, spacing));
            buttonPanel.Children.Add(CreateFormattingButton("</>", "Code (`)", InsertCode, spacing));
            buttonPanel.Children.Add(CreateFormattingButton("{ }", "Code Block (```)", InsertCodeBlock, spacing));
            buttonPanel.Children.Add(CreateFormattingButton("•", "Bullet List", InsertList, spacing));
            buttonPanel.Children.Add(CreateFormattingButton("🔗", "Link", InsertLink, spacing));
            buttonPanel.Children.Add(CreateFormattingButton("—", "Divider (---)", InsertDivider, spacing));

            _toolbarContainer.Child = buttonPanel;
            parentView.Children.Add(_toolbarContainer);
        }

        private Button CreateFormattingButton(string label, string tooltip, RoutedEventHandler action, double marginRight)
        {
            var button = new Button
            {
                Width = 40,
                Height = 30,
                Content = label,
                ToolTip = tooltip,
                Margin = new Thickness(0, 0, marginRight, 0),
                Background = Brushes.Transparent,
                BorderBrush = Brushes.Transparent,
                Foreground = Brushes.White,
                FontSize = 14,
                FontWeight = FontWeights.Medium,
                Cursor = System.Windows.Input.Cursors.Hand
            };

            button.Click += action;

            // Add hover effect
            button.MouseEnter += (s, e) =>
            {
                button.Background = new SolidColorBrush(Color.FromArgb(51, 255, 255, 255));
            };
            button.MouseLeave += (s, e) =>
            {
                button.Background = Brushes.Transparent;
            };

            return button;
        }

        public void ShowFormattingToolbar()
        {
            if (_toolbarContainer == null || _isVisible)
                return;

            _isVisible = true;
            _toolbarContainer.Visibility = Visibility.Visible;

            var fadeIn = new DoubleAnimation(0, 0.95, TimeSpan.FromMilliseconds(200));
            _toolbarContainer.BeginAnimation(UIElement.OpacityProperty, fadeIn);
        }

        public void HideFormattingToolbar()
        {
            if (_toolbarContainer == null || !_isVisible)
                return;

            _isVisible = false;

            var fadeOut = new DoubleAnimation(0.95, 0, TimeSpan.FromMilliseconds(200));
            fadeOut.Completed += (s, e) =>
            {
                _toolbarContainer.Visibility = Visibility.Hidden;
            };
            _toolbarContainer.BeginAnimation(UIElement.OpacityProperty, fadeOut);
        }

        // MARK: - Formatting Actions

        private void InsertHeading(object sender, RoutedEventArgs e)
        {
            InsertMarkdownWrapper("## ", "", "Heading");
        }

        private void InsertBold(object sender, RoutedEventArgs e)
        {
            InsertMarkdownWrapper("**", "**", "bold text");
        }

        private void InsertItalic(object sender, RoutedEventArgs e)
        {
            InsertMarkdownWrapper("*", "*", "italic text");
        }

        private void InsertCode(object sender, RoutedEventArgs e)
        {
            InsertMarkdownWrapper("`", "`", "code");
        }

        private void InsertCodeBlock(object sender, RoutedEventArgs e)
        {
            InsertMarkdownWrapper("```\n", "\n```", "code block");
        }

        private void InsertList(object sender, RoutedEventArgs e)
        {
            if (_textBox == null)
                return;

            var selectionStart = _textBox.SelectionStart;
            var selectionLength = _textBox.SelectionLength;
            var selectedText = _textBox.SelectedText;

            if (string.IsNullOrEmpty(selectedText))
                return;

            // Split by lines and add bullet to each line
            var lines = selectedText.Split(new[] { "\r\n", "\n" }, StringSplitOptions.None);
            var bulletedText = string.Join("\n", lines.Select(line =>
                string.IsNullOrEmpty(line) ? "" : $"- {line}"
            ));

            _textBox.SelectedText = bulletedText;

            // Force re-render markdown
            _textBox.Dispatcher.BeginInvoke(DispatcherPriority.Background, new Action(() =>
            {
                _renderMarkdownCallback?.Invoke();
            }));
        }

        private void InsertLink(object sender, RoutedEventArgs e)
        {
            InsertMarkdownWrapper("[", "](url)", "link text");
        }

        private void InsertDivider(object sender, RoutedEventArgs e)
        {
            if (_textBox == null)
                return;

            _textBox.SelectedText = "\n\n---\n\n";
            _renderMarkdownCallback?.Invoke();
        }

        private void InsertMarkdownWrapper(string prefix, string suffix, string placeholder)
        {
            if (_textBox == null)
                return;

            var selectedText = _textBox.SelectedText;

            // Only format if there's selected text
            if (string.IsNullOrEmpty(selectedText))
                return;

            var textToInsert = prefix + selectedText + suffix;
            _textBox.SelectedText = textToInsert;

            // Force re-render markdown
            _textBox.Dispatcher.BeginInvoke(DispatcherPriority.Background, new Action(() =>
            {
                _renderMarkdownCallback?.Invoke();
            }));
        }
    }
}

